pkg/evetestkit/utils.go

```diff
@@ -0,0 +1,475 @@
+func (node *EveNode) EveRunCommand(command string) ([]byte, error) {
+	realStdout := os.Stdout
+	r, w, err := os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+
+	os.Stdout = w
+	// unfortunately, we can't capture command return value from SSHEve
+	err = node.controller.SSHEve(command)
+	os.Stdout = realStdout
+	w.Close()
+
+	if err != nil {
+		return nil, err
+	}
+
+	out, _ := io.ReadAll(r)
+	return out, nil
+}
+
+// EveFileExists checks if a file exists on EVE node
+func (node *EveNode) EveFileExists(fileName string) (bool, error) {
+	command := fmt.Sprintf("if stat \"%s\"; then echo \"1\"; else echo \"0\"; fi", fileName)
+	out, err := node.EveRunCommand(command)
+	if err != nil {
+		return false, err
+	}
+
+	if strings.TrimSpace(string(out)) == "0" {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// EveReadFile reads a file from EVE node
+func (node *EveNode) EveReadFile(fileName string) ([]byte, error) {
+	exist, err := node.EveFileExists(fileName)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exist {
+		return nil, fmt.Errorf("file %s does not exist", fileName)
+	}
+
+	command := fmt.Sprintf("cat %s", fileName)
+	return node.EveRunCommand(command)
+}
+
+// EveDeleteFile deletes a file from EVE node
+func (node *EveNode) EveDeleteFile(fileName string) error {
+	exist, err := node.EveFileExists(fileName)
+	if err != nil {
+		return err
+	}
+
+	if !exist {
+		return nil
+	}
+
+	command := fmt.Sprintf("rm %s", fileName)
+	_, err = node.EveRunCommand(command)
+	return err
+}
+
+// WaitForAppRunningState waits for an app to start and become running on the EVE node
+func (node *EveNode) WaitForAppRunningState(appName string, timeoutSeconds uint) error {
+	start := time.Now()
+	for {
+		state, err := node.GetState(appName)
+		if err != nil {
+			return err
+		}
+
+		if strings.ToLower(state) == "running" {
+			return nil
+		}
+
+		if time.Since(start) > time.Duration(timeoutSeconds)*time.Second {
+			return fmt.Errorf("timeout waiting for app %s to start", appName)
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+}
+
+// WaitForSSH waits for the SSH connection to be established to the app VM that
+// is running on the EVE node
+func (node *EveNode) WaitForSSH(appName string, timeoutSeconds uint) error {
+	start := time.Now()
+	for {
+		_, err := node.SSHExec(appName, "echo")
+		if err == nil {
+			return nil
+		}
+
+		if time.Since(start) > time.Duration(timeoutSeconds)*time.Second {
+			return fmt.Errorf("timeout waiting for SSH connection")
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+}
+
+// StopAndRemoveApp stops and removes an app from the EVE node
+func (node *EveNode) StopAndRemoveApp(appName string) error {
+	if err := node.controller.PodStop(appName); err != nil {
+		return err
+	}
+
+	if _, err := node.controller.PodDelete(appName, true); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AddApp adds an app to the EVE node list of apps
+func (node *EveNode) AddApp(appName, SSHUser, SSHPass, SSHPort string) {
```

Comment:
What is the purpose of this function? Based on the name I would think that it would deploy a new app but it just adds some metadata to an internal list.

Reply:
this adds the app to a internal list to free us from the trouble of passing ssh user/pass/port to SSH and SCP functions.

Reply:
```
// SSHExec executes a command on the app VM running on the EVE node,
// app needs to be added using AddApp before calling this function.
func (node *EveNode) SSHExec(appName, command string) (string, error) {
```

Reply:
do you have a suggestion for a better name?

Reply:
Could this be perhaps [option(s)](https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c) for `DeployVm` function (e.g. `WithSSH(username, password, port)`)?